File: windows/src/TimeDetect.Windows/Net/DeepSeekStatusManager.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Net.Http;
using System.Runtime.CompilerServices;
using System.Threading;
using System.Threading.Tasks;

namespace TimeDetect.Net;

/// <summary>DeepSeek 官方服务状态轮询（60 秒）。</summary>
public sealed class DeepSeekStatusManager : INotifyPropertyChanged
{
    private readonly IHTTPClient _client;
    private readonly object _refreshLock = new();
    private CancellationTokenSource? _monitorCts;
    private CancellationTokenSource? _requestCts;
    private Task? _refreshTask;

    private ServiceStatusSnapshot? _snapshot;
    private bool _isRefreshing;
    private bool _isTemporarilyUnavailable;
    private DateTimeOffset? _lastUpdated;

    public static readonly Uri OfficialPageURL = new("https://status.deepseek.com");

    public ServiceStatusSnapshot? Snapshot
    {
        get => _snapshot;
        private set => SetField(ref _snapshot, value);
    }

    public bool IsRefreshing
    {
        get => _isRefreshing;
        private set => SetField(ref _isRefreshing, value);
    }

    public bool IsTemporarilyUnavailable
    {
        get => _isTemporarilyUnavailable;
        private set => SetField(ref _isTemporarilyUnavailable, value);
    }

    public DateTimeOffset? LastUpdated
    {
        get => _lastUpdated;
        private set => SetField(ref _lastUpdated, value);
    }

    public event PropertyChangedEventHandler? PropertyChanged;

    public DeepSeekStatusManager(IHTTPClient? client = null)
    {
        _client = client ?? new DefaultHTTPClient();
    }

    public void Start()
    {
        if (_monitorCts != null) return;
        Refresh();
        _monitorCts = new CancellationTokenSource();
        var ct = _monitorCts.Token;
        _ = Task.Run(async () =>
        {
            while (!ct.IsCancellationRequested)
            {
                try { await Task.Delay(TimeSpan.FromSeconds(60), ct); }
                catch (OperationCanceledException) { return; }
                Refresh();
            }
        }, ct);
    }

    public void Refresh() => _ = RefreshAsync();

    /// <summary>返回可等待的任务，便于测试确定性等待；在途重复请求会被忽略。</summary>
    public Task RefreshAsync()
    {
        lock (_refreshLock)
        {
            if (_refreshTask is { IsCompleted: false }) return Task.CompletedTask;

            var requestCts = new CancellationTokenSource();
            _requestCts = requestCts;
            _refreshTask = RefreshCoreAsync(requestCts);
            return _refreshTask;
        }
    }

    private async Task RefreshCoreAsync(CancellationTokenSource requestCts)
    {
        _isRefreshing = true;
        OnPropertyChanged(nameof(IsRefreshing));

        var ct = requestCts.Token;

        try
        {
            var value = await FetchAsync(_client, ct);
            if (ct.IsCancellationRequested) return;
            Snapshot = value;
            LastUpdated = DateTimeOffset.Now;
            IsTemporarilyUnavailable = false;
        }
        catch (OperationCanceledException) { }
        catch
        {
            // 网络、HTTP 或解码失败都不代表 DeepSeek 宕机；保留最后成功数据。
            IsTemporarilyUnavailable = true;
        }
        finally
        {
            bool ownsRefreshState;
            lock (_refreshLock)
            {
                ownsRefreshState = ReferenceEquals(_requestCts, requestCts);
                if (ownsRefreshState)
                {
                    _requestCts = null;
                    _refreshTask = null;
                }
            }
            if (ownsRefreshState)
            {
                _isRefreshing = false;
                OnPropertyChanged(nameof(IsRefreshing));
            }
        }
    }

    public void Stop()
    {
        _monitorCts?.Cancel();
        lock (_refreshLock)
        {
            _requestCts?.Cancel();
            _requestCts = null;
            _refreshTask = null;
        }
        _monitorCts = null;
        IsRefreshing = false;
    }

    public bool IsStale(DateTimeOffset? now = null)
    {
        if (_lastUpdated == null) return false;
        return (now ?? DateTimeOffset.Now) - _lastUpdated.Value > TimeSpan.FromMinutes(5);
    }

    public static async Task<ServiceStatusSnapshot> FetchAsync(
        IHTTPClient client, CancellationToken cancellationToken = default)
    {
        try
        {
            // DeepSeek 已将状态页迁移到 Flashcat。使用供应商公开 slug 可绕过
            // status.deepseek.com 当前会重置 TLS 连接的问题。
            using var request = new HttpRequestMessage(HttpMethod.Get, "https://statuspage.flashcat.cloud/deepseek");
            request.Headers.TryAddWithoutValidation("Accept", "text/html");
            var response = await client.SendAsync(request, cancellationToken);
            if (response.StatusCode < 200 || response.StatusCode >= 300)
                throw new HttpRequestException($"HTTP {(int)response.StatusCode}");
            return FlashcatParser.FlashcatPageFrom(response.Data);
        }
        catch (OperationCanceledException)
        {
            throw;
        }
        catch
        {
            // 保留迁移前 Atlassian Statuspage JSON 的兼容路径。
            using var request = new HttpRequestMessage(HttpMethod.Get, "https://status.deepseek.com/api/v2/summary.json");
            request.Headers.TryAddWithoutValidation("Accept", "application/json");
            var response = await client.SendAsync(request, cancellationToken);
            if (response.StatusCode < 200 || response.StatusCode >= 300)
                throw new HttpRequestException($"HTTP {(int)response.StatusCode}");
            return StatuspageParser.SummaryFrom(response.Data);
        }
    }

    private bool SetField<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value)) return false;
        field = value;
        OnPropertyChanged(name);
        return true;
    }

    private void OnPropertyChanged([CallerMemberName] string? name = null) =>
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
}

File: macos/Sources/App/DeepSeekStatusManager.swift
import Foundation
import Combine

@MainActor
final class DeepSeekStatusManager: ObservableObject {
    @Published private(set) var snapshot: ServiceStatusSnapshot?
    @Published private(set) var isRefreshing = false
    @Published private(set) var isTemporarilyUnavailable = false
    @Published private(set) var lastUpdated: Date?

    static let officialPageURL = URL(string: "https://status.deepseek.com")!

    private let client: HTTPClient
    private var monitorTask: Task<Void, Never>?
    private var requestTask: Task<Void, Never>?

    init(client: HTTPClient = URLSessionHTTPClient()) {
        self.client = client
    }

    func start() {
        guard monitorTask == nil else { return }
        refresh()
        monitorTask = Task { [weak self] in
            while !Task.isCancelled {
                do { try await Task.sleep(nanoseconds: 60 * 1_000_000_000) }
                catch { return }
                guard let self else { return }
                self.refresh()
            }
        }
    }

    func refresh() {
        guard !isRefreshing else { return }
        isRefreshing = true
        let client = client
        requestTask = Task { [weak self] in
            defer {
                self?.isRefreshing = false
                self?.requestTask = nil
            }
            do {
                let value = try await Self.fetch(using: client)
                guard !Task.isCancelled else { return }
                self?.snapshot = value
                self?.lastUpdated = Date()
                self?.isTemporarilyUnavailable = false
            } catch is CancellationError {
                return
            } catch {
                // 网络、HTTP 或解码失败都不代表 DeepSeek 宕机；保留最后成功数据。
                self?.isTemporarilyUnavailable = true
            }
        }
    }

    func stop() {
        monitorTask?.cancel()
        requestTask?.cancel()
        monitorTask = nil
        requestTask = nil
        isRefreshing = false
    }

    func isStale(at date: Date = Date()) -> Bool {
        guard let lastUpdated else { return false }
        return date.timeIntervalSince(lastUpdated) > 5 * 60
    }

    static func fetch(using client: HTTPClient) async throws -> ServiceStatusSnapshot {
        do {
            // DeepSeek 已将状态页迁移到 Flashcat。使用供应商公开 slug 可绕过
            // status.deepseek.com 当前会重置 TLS 连接的问题。
            var request = URLRequest(url: URL(string: "https://statuspage.flashcat.cloud/deepseek")!)
            request.httpMethod = "GET"
            request.setValue("text/html", forHTTPHeaderField: "Accept")
            let response = try await client.send(request)
            guard (200..<300).contains(response.statusCode) else {
                throw HTTPClientError.httpStatus(response.statusCode)
            }
            return try StatuspageParser.flashcatPage(from: response.data)
        } catch is CancellationError {
            throw CancellationError()
        } catch {
            // 保留迁移前 Atlassian Statuspage JSON 的兼容路径，便于官方切回
            // 或部分网络仍命中旧节点时继续工作。
            let response = try await get("summary.json", using: client)
            return try StatuspageParser.summary(from: response.data)
        }
    }

    private static func get(_ path: String, using client: HTTPClient) async throws -> HTTPResponse {
        let url = URL(string: "https://status.deepseek.com/api/v2/\(path)")!
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        let response = try await client.send(request)
        guard (200..<300).contains(response.statusCode) else {
            throw HTTPClientError.httpStatus(response.statusCode)
        }
        return response
    }
}